### PR TITLE
[READY] Blocked words tell you which ones

### PR DESCRIPTION
**Pull request by JJRcop**
## About The Pull Request

Same premise as #44712 but using a simple regex replace and also on the client side
Closes #44712

I was gonna PR it to #44712 but half way through I realized I had already replaced nearly all of bobbah's changes, and after I was finished the branches shared nothing.

## Why It's Good For The Game

~~The delta is way smaller and simpler~~
Should solve the DoS concerns without needing a limit on bad chats because ~~it's not a loop~~ it's on the client side

## Changelog
:cl: JJRcop and Bobbahbrown
tweak: When attempting to say a blocked word in character you will be notified which ones were blocked.
/:cl:

image

### CODE REVIEW

File: code/controllers/configuration/configuration.dm [hunk 1]
```diff
@@ -412,7 +413,12 @@ Example config:
...
 		if(findtextEx(line,"#",1,2))
 			continue
-		GLOB.in_character_filter += line
-
-	if(!ic_filter_regex && GLOB.in_character_filter.len)
-		ic_filter_regex = regex("\\b([jointext(GLOB.in_character_filter, "|")])\\b", "i")
+		in_character_filter += line
+
+	if(in_character_filter.len)
+		var/regex_text = 	"(\\b)([jointext(in_character_filter, "|")])(\\b)"
```
Comment (JJRcop) [inline code/controllers/configuration/configuration.dm]: `(\\b)` = `$1`, `([jointext(in_character_filter, "|")])` = `$2`, second `(\\b)` = `$3`

File: code/modules/mob/living/say.dm [hunk 2]
```diff
@@ -93,10 +93,10 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
 	if(!message || message == "")
 		return
 
-	if(GLOB.in_character_filter.len && !forced)
-		if(findtext(message, config.ic_filter_regex))
-			to_chat(src, "<span class='warning'>That message contained a word prohibited in IC chat! Consider reviewing the server rules.</span>")
-			return
+	if(!forced && config.ic_filter_regex && findtext(message, config.ic_filter_regex))
+		var/list/words = replacetext(message, config.ic_filter_regex_G, "$1<span class='boldwarning'>$2</span>$3")
```
Comment (JJRcop) [inline code/modules/mob/living/say.dm]: See https://github.com/tgstation/tgstation/pull/44728#discussion_r296431139

**Comment by bobbahbrown** [comment]
I like this, now I see what you meant! I will close mine this one makes more sense

**Comment by WJohn** [comment]
so uh, what's the list of blocked words? is this something that's going to be set up and expanded by the admins of the server in question, then?

**Comment by JJRcop** [comment]
That's correct, see #44668

**Comment by JJRcop** [comment]
I'd like to move the replacetext to the client, looking into that

**Comment by JJRcop** [comment]
The warning message that shows you which words are blocked is now handled by goonchat on the client side

**Comment by ike709** [comment]
@JJRcop There was a bit of an oversight with the original implementation, discovered by @nervere.

Several regex symbols (`()^` are confirmed, but probably all of them), break the filter if they are included in the config file. For example, filtering `:(` or `:)`. Could you either add a warning to the file, or even better, skip lines containing special characters or replacetext them to include an escape char?

If you don't want to, I will after this PR is merged.

**Comment by JJRcop** [comment]
I wonder if I want to fully support regex in the file or escape all of them. The normal fix for that sort of thing is using :\), with \) as an escaped paren.

**Comment by optimumtact** [comment]
escape them please, allowing arbitrary regex is asking for a DOS vector really

**Comment by optimumtact** [comment]
In the original matching that is, not your goonchat change

**Comment by JJRcop** [comment]
Tried but wasn't able to fix the emoticon issue at this time

**Comment by JJRcop** [comment]
I fixed the issue @ike709 mentioned. The client's word replace doesn't work well with non-alphanums, or any regex that might be affected by sanitization. Still works normally though.

it is an image

**Comment by AffectedArc07** [comment]
Can I ask why we switched from using a proper global to `var/list/`

**Comment by JJRcop** [comment]
You certainly can. It was not being used anywhere other than to generate the regex so I pulled it out of global. If you need it to be global just put it back.

**Comment by SouDescolado** [comment]
Could someone with NEET quirk have a way to bypass this block? I know some people in real life that say "lol"

**Comment by optimumtact** [comment]
The filter was in an entirely different PR and has no bearing on this one, take your questions about the filters themselves to the policy forum.

**Pull request merged**